### ISSUE TITLE
Small window to merge auto-update PRs when CI is slow and dependencies updating rapidly

### ISSUE THREAD

**Issue by maririos** [dotnet/buildtools#1902]
@dagood commented on [Fri Jan 26 2018](https://github.com/dotnet/core-eng/issues/2519)

The goal is to make auto-PRs more mergeable. Wanted to write down ideas after briefly talking about it with @MattGal. ~~Summary: I don't have any ideas that seem feasible short-term.~~

### (New) Stable/Volatile PR pair

> [@chcosta] ([comment](https://github.com/dotnet/core-eng/issues/2519#issuecomment-362421062))
> ### 2 Active PR's
>
> Another option is to permit multiple PR's, but capped at two.  This is sort of a stable / volatile model.  The algorithm would look like this.
> 
> 1. If there are no open PR's, open one.
> 2. If there is one open PR, open a second PR.
> 3. If there are two open PR's, update the second PR.
> 
> With this model, you have a stable PR which does not change, and then a volatile PR which is collecting updates.  As soon as the first PR is closed, the second PR becomes stable.
> 
> An additional note from Davis on this model...
> 
> > The auto-PR submitter can modify the description to send that message, actually.
> Even link the two together.
> 
> Davis estimates that this model would be "less than a day's work".

I really like this idea. It solves the merging problem, doesn't require any novel infrastructure, and it lets devs continue to see "bleeding-edge" auto-update commits as they come in without affecting mergeability. Money is still burned by the volatile PR, but that's not what this issue's about.

---

Old ideas:

### Rate limit auto-PRs
We could have the auto-update tooling simply not update the PR if it was updated recently. (Better: if CI is in progress, or only finished recently.) Straightforward, but:

 * You lose updates. If a bunch of version changes happen at once, and they're discarded, you have to wait until the next version change to get any auto-update. The repo would be left cold with old versions if the changes stop completely.
 * There's still a hard cutoff: you have to check the auto-PR while it's in the rate limit window.

### Make it easy to see if a previous update ended up with green CI and to merge it
VersionTools already posts a comment telling you what the old update commit was and the status of CI when it was discarded. This is a step towards being able to merge an old upgrade to at least get something new. But:

 * Currently, there's no clear way to actually fetch that old update commit, because it's not in any branch.
 * The CI status is "frozen" at the time the comment is made. You can click into each "in progress" Details link to see current status, but this is tedious and error-prone.
 * Merging an old update commit guarantees a merge conflict with the new update commits. You lose those updates, just like the approach above. (Maestro/VersionTools doesn't have any PR activity hooks or fancy UI, so there's no way to tell it to create a new update on demand.)

### Auto-merge PRs
This is the most reasonable short-term thing I can think of. If we merge PRs as soon as CI is green, we don't need a large window. But:

 * If CI *never* completes before the next auto-update commit, we haven't improved anything.
 * Security concerns. https://github.com/dotnet/core-eng/issues/939 for details.

### Only create new commits/PRs on demand
One big thing that would open up a bunch of possibilities is having a way to trigger things when auto-update PRs are merged or closed. It might make sense for Maestro to handle these events, but it would need to be significantly smarter.

---

@MattGal commented on [Fri Jan 26 2018](https://github.com/dotnet/core-eng/issues/2519#issuecomment-360915505)

Excellent writeup, you rock Davis!  @joshfree FYI

@ChadNedzlek we need to explore how much time (%-wise) is spent in a CI run on Kusto processing and see if we can make a workaround for this.  In simple calculations (the telemetry for this is stored in more than one place unfortunately) it looks like a CI run can easily spend > 50% of test execution time waiting on Kusto ingestion

---

@ChadNedzlek commented on [Fri Jan 26 2018](https://github.com/dotnet/core-eng/issues/2519#issuecomment-360916049)

We have metrics for all this as soon as we deploy. We'll have metrics for how long it took for the SQL processor to see events, then how long it took the Kusto processor to handle the "finished" notification, and we'll have an end to end time from "event was generated" to "it is available in Kusto"

---

@joshfree commented on [Fri Jan 26 2018](https://github.com/dotnet/core-eng/issues/2519#issuecomment-360916584)

> If CI never completes before the next auto-update commit, we haven't improved anything.

Yes, this is the situation I've been observing the past week.  I wonder how much $$ we may be burning this way. 😄 

---

@dagood commented on [Fri Jan 26 2018](https://github.com/dotnet/core-eng/issues/2519#issuecomment-360919232)

> having a way to trigger things when auto-update PRs are merged or closed.

Briefly expanding on a big idea I have with this (would avoid a bunch of 🔥💵💵💵🔥): we could change to *never* submit new commits when there's an existing upgrade PR. The bot would comment on the PR about the new version available for awareness, but wouldn't push a new update commit anywhere. If you merge or close the PR, the bot generates a new PR with the latest versions. The more I think about this, the more I like it.

---

@maririos commented on [Thu Feb 01 2018](https://github.com/dotnet/core-eng/issues/2519#issuecomment-362416480)

@chcosta could you please take a look at this to see what options we could have to solve this?

---

@chcosta commented on [Thu Feb 01 2018](https://github.com/dotnet/core-eng/issues/2519#issuecomment-362421062)

### 2 Active PR's

Another option is to permit multiple PR's, but capped at two.  This is sort of a stable / volatile model.  The algorithm would look like this.

1. If there are no open PR's, open one.
2. If there is one open PR, open a second PR.
3. If there are two open PR's, update the second PR.

With this model, you have a stable PR which does not change, and then a volatile PR which is collecting updates.  As soon as the first PR is closed, the second PR becomes stable.

An additional note from Davis on this model...

> The auto-PR submitter can modify the description to send that message, actually.
Even link the two together.

Davis estimates that this model would be "less than a day's work".

**Comment by eerhardt** [comment]
Another idea that should be pretty cheap to implement, but doesn't completely fix all concerns, is to create a new Label that the auto-PRs look for.  Let's call this Label "Maestro-Stop-Updating".

When CI is lagging, and I want this change in and don't want Maestro to send any more updates, I add the Label "Maestro-Stop-Updating" to the PR.  If a new update comes along, the auto-PR tool sees the Label is applied, and instead of updating the PR, it does what it does today when Maestro-bot isn't the HEAD author of the PR.  It simply writes, "Hey I would have updated this PR, but you told me not to".

This gives humans the chance to see when the CI is green, and merge it when they want.  And it stops Maestro from updating the PR.  When the next update comes out after the merge, it will do all the new updates.  This gets us in the same situation as when a dev updates an auto-PR.  But we create new PRs often enough that this shouldn't be a huge issue in `master` branches.  I get new updates every few hours.  And if someone doesn't like/want this functionality, they don't need to set the Label ever and their current flow won't change.

**Comment by dagood** [comment]
Starting with BuildTools `2.2.0-preview1-02725-02`, the auto-updater won't push new commits to PRs that have the `Maestro-Stop-Updating` label (case-insensitive).

**Comment by ViktorHofer** [comment]
// Auto-generated message

Going forward, the .NET team is using https://github.com/dotnet/arcade to develop the code and issues formerly in this repository. Feel free to reopen/move this issue if it still applies to servicing branches in this repository, or source code which now resides in the Arcade repository.
